src/Collection.php

```diff
@@ -810,6 +811,28 @@ public function listIndexes(array $options = [])
     }
 
     /**
+     * Runs map-reduce aggregation operations as defined by the $map and $reduce parameters.
+     *
+     * @see MapReduce::__construct() for supported options
+     * @see http://docs.mongodb.org/manual/reference/command/mapReduce/
+     * @param JavascriptInterface $map            Map function
+     * @param JavascriptInterface $reduce         Reduce function
+     * @param string|array|object $out            Output specification
+     * @param array               $options        Command options
+     * @return MapReduceResult
+     * @throws UnsupportedException if options are not supported by the selected server
+     * @throws InvalidArgumentException for parameter/option parsing errors
+     * @throws DriverRuntimeException for other driver errors (e.g. connection errors)
+     */
+    public function mapReduce(JavascriptInterface $map, JavascriptInterface $reduce, $out, array $options = [])
+    {
+        $operation = new MapReduce($this->databaseName, $this->collectionName, $map, $reduce, $out, $options);
+        $server = $this->manager->selectServer(new ReadPreference(ReadPreference::RP_PRIMARY));
```

Comment:
The read preference is configurable, but we will want to coerce a primary read preference if the out option is not inline. That will be similar to how `aggregate()` checks if the `$out` pipeline operator is being used and, if so, forces a primary read preference. After we establish the read preference, we then use that to select a server upon which the operation class will be executed.

Since the out option can be a document (i.e. PHP array or object), I'd suggest looking at `is_last_pipeline_operator_out()` in `functions.php` for a resilient way to check. Something like the following should work:

```
// In this case, $out is probably a string (i.e. collection name)
if ( ! is_array($out) && ! is_object($out)) {
    return false;
}

// Cast a possible object to an array so we can conveniently check the first field name
$out = (array) $out;

return key($out) === 'inline';
```

Additionally, the type map, read concern, and write concern can all be inherited from the collection (if unspecified in the options). See `aggregate()` for an example of how we merge those in. Notably, read concern depends on the server's wire version, so we only inject a default value if we know it's supported.